File: notification_system_impl_python/providers/smtp_provider.py
import smtplib
import ssl
import json
import os
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
from typing import Dict, Any, Optional
from providers.base_provider import NotificationProvider
from dotenv import load_dotenv

load_dotenv()



class SMTPProvider(NotificationProvider):
    def __init__(self, host: str, port: int, username: str, password: str, from_email: str = None, use_tls: bool = True):
        self.host = host
        self.port = port
        self.username = username
        self.password = password
        self.from_email = from_email or username
        self.use_tls = use_tls

    def provider_name(self) -> str:
        return "smtp"

    def send(self, notification) -> Dict[str, Any]:
        try:
            payload = json.loads(notification.payload)

            to_email = payload.get('to')
            subject = payload.get('subject', 'Notification')
            body = payload.get('body', '')
            from_email = payload.get('from', self.from_email)

            if not to_email:
                return {'success': False, 'message': 'Missing "to" field in payload'}

            if not body:
                return {'success': False, 'message': 'Missing "body" field in payload'}

            message = MIMEMultipart("alternative")
            message["Subject"] = subject
            message["From"] = from_email
            message["To"] = to_email

            html_part = MIMEText(body, "html")
            message.attach(html_part)
            context = ssl.create_default_context()
            if self.use_tls:
                with smtplib.SMTP(self.host, self.port) as server:
                    server.starttls(context=context)
                    server.login(self.username, self.password)
                    server.send_message(message)
            else:
                with smtplib.SMTP_SSL(self.host, self.port, context=context) as server:
                    server.login(self.username, self.password)
                    server.send_message(message)

            return {
                'success': True,
                'message': f'Email sent via SMTP to {to_email}',
                'response': {'to': to_email, 'subject': subject}
            }

        except Exception as e:
            return {
                'success': False,
                'message': f'SMTP send failed: {str(e)}'
            }


class GmailProvider(SMTPProvider):

    def __init__(self, email: str, app_password: str):
        super().__init__(
            host="smtp.gmail.com",
            port=587,
            username=email,
            password=app_password,
            from_email=email,
            use_tls=True
        )

    def provider_name(self) -> str:
        return "gmail"


class OutlookProvider(SMTPProvider):

    def __init__(self, email: str, password: str):
        super().__init__(
            host="smtp-mail.outlook.com",
            port=587,
            username=email,
            password=password,
            from_email=email,
            use_tls=True
        )

    def provider_name(self) -> str:
        return "outlook"


def load_smtp_provider_from_env() -> Optional[SMTPProvider]:
    provider = os.getenv("SMTP_PROVIDER", "").lower()

    if provider == "gmail":
        email = os.getenv("GMAIL_EMAIL")
        password = os.getenv("GMAIL_APP_PASSWORD")
        if email and password:
            return GmailProvider(email, password)

    elif provider == "outlook":
        email = os.getenv("OUTLOOK_EMAIL")
        password = os.getenv("OUTLOOK_PASSWORD")
        if email and password:
            return OutlookProvider(email, password)

    elif provider == "custom":
        return SMTPProvider(
            host=os.getenv("SMTP_HOST", "localhost"),
            port=int(os.getenv("SMTP_PORT", "587")),
            username=os.getenv("SMTP_USERNAME", ""),
            password=os.getenv("SMTP_PASSWORD", ""),
            from_email=os.getenv("SMTP_FROM_EMAIL"),
            use_tls=os.getenv("SMTP_USE_TLS", "true").lower() == "true"
        )

    return None

File: notification_system_impl_go/core/providers/smtp_provider.go
package providers

import (
	"context"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"net/smtp"
	"os"
	"time"

	"notification_system/models"
)

type SMTPProvider struct {
	Host     string
	Port     string
	Username string
	Password string
	From     string
}

func NewSMTPProvider(host, port, username, password, from string) *SMTPProvider {
	return &SMTPProvider{
		Host:     host,
		Port:     port,
		Username: username,
		Password: password,
		From:     from,
	}
}

func NewGmailProvider(email, password string) *SMTPProvider {
	return &SMTPProvider{
		Host:     "smtp.gmail.com",
		Port:     "587",
		Username: email,
		Password: password,
		From:     email,
	}
}

func NewOutlookProvider(email, password string) *SMTPProvider {
	return &SMTPProvider{
		Host:     "smtp-mail.outlook.com",
		Port:     "587",
		Username: email,
		Password: password,
		From:     email,
	}
}

func (s *SMTPProvider) ProviderName() string {
	return "smtp"
}

func (s *SMTPProvider) Send(ctx context.Context, notification *models.Notification) error {
	var payload map[string]interface{}
	if err := json.Unmarshal([]byte(notification.Payload), &payload); err != nil {
		return fmt.Errorf("invalid payload: %w", err)
	}

	to, ok := payload["to"].(string)
	if !ok {
		return fmt.Errorf("missing 'to' field in payload")
	}

	subject, ok := payload["subject"].(string)
	if !ok {
		subject = "Notification"
	}

	body, ok := payload["body"].(string)
	if !ok {
		return fmt.Errorf("missing 'body' field in payload")
	}

	from := s.From
	if payloadFrom, ok := payload["from"].(string); ok && payloadFrom != "" {
		from = payloadFrom
	}

	message := fmt.Sprintf("From: %s\r\n", from)
	message += fmt.Sprintf("To: %s\r\n", to)
	message += fmt.Sprintf("Subject: %s\r\n", subject)
	message += fmt.Sprintf("MIME-Version: 1.0\r\n")
	message += fmt.Sprintf("Content-Type: text/html; charset=UTF-8\r\n")
	message += fmt.Sprintf("\r\n%s\r\n", body)

	auth := smtp.PlainAuth("", s.Username, s.Password, s.Host)

	addr := fmt.Sprintf("%s:%s", s.Host, s.Port)

	tlsConfig := &tls.Config{
		InsecureSkipVerify: false,
		ServerName:         s.Host,
	}

	conn, err := smtp.Dial(addr)
	if err != nil {
		return fmt.Errorf("failed to connect to SMTP server: %w", err)
	}
	defer conn.Close()

	if err := conn.StartTLS(tlsConfig); err != nil {
		return fmt.Errorf("failed to start TLS: %w", err)
	}

	if err := conn.Auth(auth); err != nil {
		return fmt.Errorf("SMTP authentication failed: %w", err)
	}

	if err := conn.Mail(from); err != nil {
		return fmt.Errorf("failed to set sender: %w", err)
	}

	if err := conn.Rcpt(to); err != nil {
		return fmt.Errorf("failed to set recipient: %w", err)
	}

	writer, err := conn.Data()
	if err != nil {
		return fmt.Errorf("failed to get data writer: %w", err)
	}

	_, err = writer.Write([]byte(message))
	if err != nil {
		return fmt.Errorf("failed to write message: %w", err)
	}

	if err := writer.Close(); err != nil {
		return fmt.Errorf("failed to close writer: %w", err)
	}

	if err := conn.Quit(); err != nil {
		return fmt.Errorf("failed to quit connection: %w", err)
	}

	notification.Status = models.StatusSent
	now := time.Now().UnixMilli()
	notification.SentAt = &now
	response := fmt.Sprintf("Email sent via SMTP to %s", to)
	notification.ProviderResponse = &response

	return nil
}

func LoadSMTPProviderFromEnv() *SMTPProvider {
	provider := os.Getenv("SMTP_PROVIDER")

	if provider == "gmail" {
		email := os.Getenv("GMAIL_EMAIL")
		password := os.Getenv("GMAIL_APP_PASSWORD")
		return NewGmailProvider(email, password)
	}

	if provider == "outlook" {
		email := os.Getenv("OUTLOOK_EMAIL")
		password := os.Getenv("OUTLOOK_PASSWORD")
		return NewOutlookProvider(email, password)
	}

	return NewSMTPProvider(
		os.Getenv("SMTP_HOST"),
		os.Getenv("SMTP_PORT"),
		os.Getenv("SMTP_USERNAME"),
		os.Getenv("SMTP_PASSWORD"),
		os.Getenv("SMTP_FROM_EMAIL"),
	)
}
